src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -119,71 +119,71 @@ SymbolKind.NetModule or
                 CompilationAndGeneratorDriverTranslationAction? translate)
             {
-                var forkedTrackerState = ForkTrackerState(oldProjectState: this.ProjectState, ReadState(), translate);
+                var oldProjectState = this.ProjectState;
+                var forkedTrackerState = ForkTrackerState();
 
                 // We should never fork into a FinalCompilationTrackerState.  We must always be at some state prior to
                 // it since some change has happened, and we may now need to run generators.
                 Contract.ThrowIfTrue(forkedTrackerState is FinalCompilationTrackerState);
+                Contract.ThrowIfFalse(forkedTrackerState is null or InProgressState);
                 return new CompilationTracker(
                     newProjectState,
                     forkedTrackerState,
                     this.SkeletonReferenceCache.Clone());
 
-                static CompilationTrackerState? ForkTrackerState(
-                    ProjectState oldProjectState,
-                    CompilationTrackerState? state,
-                    CompilationAndGeneratorDriverTranslationAction? translate)
+                CompilationTrackerState? ForkTrackerState()
                 {
+                    var state = this.ReadState();
                     if (state is null)
                         return null;
 
-                    // Determine the old/stale compilation to help seed the non-final state with.
-                    var staleCompilationWithGeneratedDocuments = state switch
+                    var (compilationWithoutGeneratedDocuments, staleCompilationWithGeneratedDocuments) = state switch
                     {
-                        FinalCompilationTrackerState finalState => finalState.FinalCompilationWithGeneratedDocuments,
-                        NonFinalWithCompilationTrackerState nonFinalState => nonFinalState.StaleCompilationWithGeneratedDocuments,
+                        InProgressState inProgressState => (inProgressState.CompilationWithoutGeneratedDocuments, inProgressState.StaleCompilationWithGeneratedDocuments),
+                        FinalCompilationTrackerState finalState => (finalState.CompilationWithoutGeneratedDocuments, finalState.FinalCompilationWithGeneratedDocuments),
                         _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
                     };
 
-                    var intermediateProjects = UpdateIntermediateProjects(oldProjectState, state, translate);
+                    var finalSteps = UpdatePendingTranslationSteps(
+                        state switch
+                        {
+                            InProgressState { PendingTranslationSteps: var pendingTranslationSteps } => pendingTranslationSteps,
+                            FinalCompilationTrackerState => [],
+                            _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
+                        });
```

Comment:
Since you've already got a switch you could try to sneak this switch into the tuple you've already got above.